### ISSUE TITLE
[RFC] Extending namings to allow utf-8 symbols.

### ISSUE THREAD

**Issue by mhermier** [wren-lang/wren#891]
As of now, names are limited to ascii-7 charset. It would be nice to also allow utf-8 symbols, to have Greek annotation, possibly localized names for users... It would probably requires an external lib to get characters classes, but maybe it is possible to allow some of the symbols for cheaper ?

**Comment by PureFox48** [comment]
Well a lot of 'newer' languages support the use of non-ASCII characters within identifiers and I have a lot of sympathy with this as it must be frustrating for people whose first language isn't English not to be able to write their identifiers in their native language.

In practice there are all sorts of difficulties - should one treat an accented character as being equivalent to the corresponding unaccented character plus the accent, what to do about capitalization (which matters for the likes of Go and Wren) where languages have no distinction between upper and lower case letters, the difficulty for most Westerners of reading Chinese or Japanese characters - and so forth. The Go team (which includes the inventors of UTF-8) have been agonizing over these sorts of questions for years.

Wren is a simple language and my personal view is that, if we do decide to go beyond the ASCII character set for identifiers, then we should look for a simple solution that doesn't need to use an external library to decide whether something is an acceptable character or not.

In my own libraries, I have a sort of 'half-way house' where my capitalization, pattern matching functions etc. deal with characters in the [IS0-8859-1](https://en.wikipedia.org/wiki/ISO/IEC_8859-1) range (code-points 0-255 in Unicode). This is enough to cover all the major Western European languages (apart from a few rare characters) but doesn't unfortunately cover the Greek alphabet. However, it does have the merit of being easy to program without the need for a table (all the letters are in blocks) and so I'm reasonably happy with it. Of course, I'd be less happy if I were from a country whose characters were not supported!

**Comment by mhermier** [comment]
I agree about the difficulties, it not meant to be for today, but more to
have a discussion entry to refer to.

About lowering accented symbols, I don't think it is a wise decision, like
ignoring case. It has its reasons and usage in the past, but with the
improvement of editors, we are pushing in fact what is an editor problem to
the compiler. To make an analogy, if you need to make square holes and only
can drill round holes, you can either change of tool or change the
standard. In the past changing the tool was luxury, but today it is less
more true.

You are right about difficulties of reading of foreign languages, this can
be solved by some solutions. I can think of 2 for now: localisation of
symbols, aliasing of symbols. While basically the same thing, I prefer
aliasing, because it is more opt-in and the language does not magically
change based on your locale.

**Comment by clsource** [comment]
I find it useful for having variable and method names with non english chars like `Ñ`.
perhaps using a special header comment similar to python https://docs.python.org/3/reference/lexical_analysis.html#encoding-declarations

`# -*- coding: utf-8 -*-`

to declare which symbols would be available as valid variable and method names. 
I think the technique of symbol alias would be the best choice for replacing utf8 chars with ascii compatible ones.
